library/system/io.lean

```diff
@@ -68,6 +91,14 @@ interface.iterate e α a f
 def forever {e} (a : io_core e unit) : io_core e unit :=
 iterate () $ λ _, a >> return (some ())
 
+def catch {e₁ e₂ α} (a : io_core e₁ α) (b : e₁ → io_core e₂ α) : io_core e₂ α :=
+interface.catch e₁ e₂ α a b
+
+instance : alternative io :=
+{ interface.monad _ with
+  orelse := λ _ a b, catch a (λ _, b),
+  failure := λ _, io.fail "failure" }
+
```

Comment:
@gebner I'm not sure whether `io` should be an instance of `alternative` or not.
Right now, it is not an issue because we don't have laws for `alternative`, nor `laws` showing how `failure` interacts with `bind`. We will eventually need laws for `fail` and `catch` too.
This is messy in Haskell (@Kha posted a link to the Haskell documentation where they discuss the `alternative` laws). That being said, I think it would be nice if we had laws such as `a >> failure = failure` for the `alternative` type class, which are not true for `io`. 
Anyway, I'm merging this PR as-is. We can decide this issue later.